File: AgentsMonitor/Models/Session.swift
import Foundation
import Darwin

struct Session: Identifiable, Hashable {
    let id: UUID
    var name: String
    var status: SessionStatus
    var agentType: AgentType
    var startedAt: Date
    var endedAt: Date?
    var messages: [Message]
    var toolCalls: [ToolCall]
    var metrics: SessionMetrics
    var workingDirectory: URL?
    var processId: Int32?
    var errorMessage: String?
    var isExternalProcess: Bool
    var isFullyLoaded: Bool
    var terminalOutput: Data?
    var exitCode: Int32?

    // Claude session fields
    var jsonlPath: String?
    var projectPath: String?
    var gitBranch: String?
    var firstPrompt: String?
    var sessionSummary: String?
    var isSidechain: Bool = false
    var fileMtime: Int64 = 0

    init(
        id: UUID = UUID(),
        name: String,
        status: SessionStatus = .running,
        agentType: AgentType = .claudeCode,
        startedAt: Date = Date(),
        endedAt: Date? = nil,
        messages: [Message] = [],
        toolCalls: [ToolCall] = [],
        metrics: SessionMetrics = SessionMetrics(),
        workingDirectory: URL? = nil,
        processId: Int32? = nil,
        errorMessage: String? = nil,
        isExternalProcess: Bool = false,
        isFullyLoaded: Bool = true,
        terminalOutput: Data? = nil,
        exitCode: Int32? = nil,
        jsonlPath: String? = nil,
        projectPath: String? = nil,
        gitBranch: String? = nil,
        firstPrompt: String? = nil,
        sessionSummary: String? = nil,
        isSidechain: Bool = false,
        fileMtime: Int64 = 0
    ) {
        self.id = id
        self.name = name
        self.status = status
        self.agentType = agentType
        self.startedAt = startedAt
        self.endedAt = endedAt
        self.messages = messages
        self.toolCalls = toolCalls
        self.metrics = metrics
        self.workingDirectory = workingDirectory
        self.processId = processId
        self.errorMessage = errorMessage
        self.isExternalProcess = isExternalProcess
        self.isFullyLoaded = isFullyLoaded
        self.terminalOutput = terminalOutput
        self.exitCode = exitCode
        self.jsonlPath = jsonlPath
        self.projectPath = projectPath
        self.gitBranch = gitBranch
        self.firstPrompt = firstPrompt
        self.sessionSummary = sessionSummary
        self.isSidechain = isSidechain
        self.fileMtime = fileMtime
    }

    func duration(asOf date: Date) -> TimeInterval {
        let end = endedAt ?? date
        return end.timeIntervalSince(startedAt)
    }

    func formattedDuration(asOf date: Date) -> String {
        let interval = duration(asOf: date)
        let formatter = DateComponentsFormatter()
        formatter.unitsStyle = .abbreviated
        if interval < 60 {
            formatter.allowedUnits = [.second]
        } else {
            formatter.allowedUnits = [.day, .hour, .minute]
        }
        return formatter.string(from: interval) ?? "0s"
    }

    var relativeTimeString: String {
        let interval = Date().timeIntervalSince(startedAt)
        if interval < 60 { return "just now" }
        let formatter = DateComponentsFormatter()
        formatter.unitsStyle = .abbreviated
        formatter.maximumUnitCount = 2
        if interval < 3600 {
            formatter.allowedUnits = [.minute]
        } else if interval < 86400 {
            formatter.allowedUnits = [.hour, .minute]
        } else {
            formatter.allowedUnits = [.day, .hour]
        }
        guard let formatted = formatter.string(from: interval) else { return "just now" }
        return "\(formatted) ago"
    }

    var shortProjectName: String? {
        guard let path = projectPath else { return nil }
        let components = path.split(separator: "/")
        if components.count >= 2 {
            return components.suffix(2).joined(separator: "/")
        }
        return components.last.map(String.init)
    }

    var duration: TimeInterval {
        duration(asOf: Date())
    }

    var formattedDuration: String {
        formattedDuration(asOf: Date())
    }

    static func == (lhs: Session, rhs: Session) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

enum SessionStatus: String, CaseIterable, Codable {
    case running = "Running"
    case paused = "Paused"
    case completed = "Completed"
    case failed = "Failed"
    case waiting = "Waiting"
    case cancelled = "Cancelled"

    var color: String {
        switch self {
        case .running: return "green"
        case .paused: return "yellow"
        case .completed: return "blue"
        case .failed: return "red"
        case .waiting: return "orange"
        case .cancelled: return "gray"
        }
    }

    var icon: String {
        switch self {
        case .running: return "play.circle.fill"
        case .paused: return "pause.circle.fill"
        case .completed: return "checkmark.circle.fill"
        case .failed: return "xmark.circle.fill"
        case .waiting: return "clock.fill"
        case .cancelled: return "stop.circle.fill"
        }
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let raw = try container.decode(String.self)
        let normalized = Self.normalizedDecodeValue(raw)

        switch normalized {
        case "running":
            self = .running
        case "paused":
            self = .paused
        case "completed":
            self = .completed
        case "failed":
            self = .failed
        case "waiting":
            self = .waiting
        case "cancelled", "canceled":
            self = .cancelled
        default:
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Invalid SessionStatus value: \(raw)"
            )
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }

    private static func normalizedDecodeValue(_ value: String) -> String {
        value
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: "_", with: "")
            .replacingOccurrences(of: "-", with: "")
    }
}

enum AgentType: String, CaseIterable, Codable {
    case claudeCode = "Claude Code"

    var icon: String { "brain" }
    var displayName: String { rawValue }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let raw = try container.decode(String.self)
        let normalized = raw
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: "_", with: "")
            .replacingOccurrences(of: "-", with: "")

        switch normalized {
        case "claudecode", "claude", "anthropicclaude", "claudecli", "claudecodecli", "claudecodeagent":
            self = .claudeCode
        default:
            // Default to claudeCode for unknown types
            self = .claudeCode
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

struct SessionMetrics: Hashable, Codable {
    static let defaultContextWindowMax = 200_000

    var totalTokens: Int
    var inputTokens: Int
    var outputTokens: Int
    var toolCallCount: Int
    var errorCount: Int
    var apiCalls: Int
    var cacheReadTokens: Int
    var cacheWriteTokens: Int
    var contextWindowMax: Int
    var cost: Double
    var modelName: String

    init(
        totalTokens: Int = 0,
        inputTokens: Int = 0,
        outputTokens: Int = 0,
        toolCallCount: Int = 0,
        errorCount: Int = 0,
        apiCalls: Int = 0,
        cacheReadTokens: Int = 0,
        cacheWriteTokens: Int = 0,
        contextWindowMax: Int = defaultContextWindowMax,
        cost: Double = 0.0,
        modelName: String = ""
    ) {
        self.totalTokens = totalTokens
        self.inputTokens = inputTokens
        self.outputTokens = outputTokens
        self.toolCallCount = toolCallCount
        self.errorCount = errorCount
        self.apiCalls = apiCalls
        self.cacheReadTokens = cacheReadTokens
        self.cacheWriteTokens = cacheWriteTokens
        self.contextWindowMax = contextWindowMax
        self.cost = cost
        self.modelName = modelName
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        totalTokens = try container.decode(Int.self, forKey: .totalTokens)
        inputTokens = try container.decode(Int.self, forKey: .inputTokens)
        outputTokens = try container.decode(Int.self, forKey: .outputTokens)
        toolCallCount = try container.decode(Int.self, forKey: .toolCallCount)
        errorCount = try container.decode(Int.self, forKey: .errorCount)
        apiCalls = try container.decode(Int.self, forKey: .apiCalls)
        cacheReadTokens = (try? container.decodeIfPresent(Int.self, forKey: .cacheReadTokens)) ?? 0
        cacheWriteTokens = (try? container.decodeIfPresent(Int.self, forKey: .cacheWriteTokens)) ?? 0
        contextWindowMax = (try? container.decodeIfPresent(Int.self, forKey: .contextWindowMax)) ?? Self.defaultContextWindowMax
        cost = (try? container.decodeIfPresent(Double.self, forKey: .cost)) ?? 0.0
        modelName = (try? container.decodeIfPresent(String.self, forKey: .modelName)) ?? ""
    }

    enum CodingKeys: String, CodingKey {
        case totalTokens, inputTokens, outputTokens
        case toolCallCount, errorCount, apiCalls
        case cacheReadTokens, cacheWriteTokens
        case contextWindowMax, cost, modelName
    }

    var contextWindowUsage: Double {
        guard contextWindowMax > 0 else { return 0 }
        return min(max(Double(totalTokens) / Double(contextWindowMax), 0), 1.0)
    }

    var formattedWindowMax: String {
        contextWindowMax >= 1_000 ? "\(contextWindowMax / 1_000)K" : "\(contextWindowMax)"
    }

    var formattedRemaining: String {
        let remaining = max(contextWindowMax - totalTokens, 0)
        if remaining >= 1_000 {
            return String(format: "%.1fK", Double(remaining) / 1_000)
        }
        return "\(remaining)"
    }

    var formattedContextWindow: String {
        "\(formattedTokens) / \(formattedWindowMax)"
    }

    var formattedTokens: String {
        let total = totalTokens
        if total >= 1_000_000 {
            return String(format: "%.1fM", Double(total) / 1_000_000)
        } else if total >= 1_000 {
            return String(format: "%.1fK", Double(total) / 1_000)
        } else {
            return "\(total)"
        }
    }

    var formattedCost: String {
        cost > 0 ? String(format: "$%.4f", cost) : "--"
    }
}

File: agentsmonitor-tauri/src-tauri/src/models/session.rs
use base64::{engine::general_purpose::STANDARD, Engine};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use uuid::Uuid;

use super::{Message, ToolCall};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Session {
    pub id: Uuid,
    pub name: String,
    pub status: SessionStatus,
    pub agent_type: AgentType,
    pub started_at: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ended_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub messages: Vec<Message>,
    #[serde(default)]
    pub tool_calls: Vec<ToolCall>,
    #[serde(default)]
    pub metrics: SessionMetrics,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub working_directory: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub process_id: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_message: Option<String>,
    #[serde(default)]
    pub is_external_process: bool,
    #[serde(default = "default_true")]
    pub is_fully_loaded: bool,
    #[serde(
        skip_serializing_if = "Option::is_none",
        serialize_with = "serialize_terminal_output",
        deserialize_with = "deserialize_terminal_output",
        default
    )]
    pub terminal_output: Option<Vec<u8>>,
}

fn serialize_terminal_output<S>(data: &Option<Vec<u8>>, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    match data {
        Some(bytes) => serializer.serialize_str(&STANDARD.encode(bytes)),
        None => serializer.serialize_none(),
    }
}

fn deserialize_terminal_output<'de, D>(deserializer: D) -> Result<Option<Vec<u8>>, D::Error>
where
    D: Deserializer<'de>,
{
    let opt: Option<String> = Option::deserialize(deserializer)?;
    match opt {
        Some(s) => STANDARD
            .decode(&s)
            .map(Some)
            .map_err(serde::de::Error::custom),
        None => Ok(None),
    }
}

fn default_true() -> bool {
    true
}

impl Session {
    pub fn new(name: String, agent_type: AgentType) -> Self {
        Self {
            id: Uuid::new_v4(),
            name,
            status: SessionStatus::Running,
            agent_type,
            started_at: Utc::now(),
            ended_at: None,
            messages: Vec::new(),
            tool_calls: Vec::new(),
            metrics: SessionMetrics::default(),
            working_directory: None,
            process_id: None,
            error_message: None,
            is_external_process: false,
            is_fully_loaded: true,
            terminal_output: None,
        }
    }

    pub fn duration_secs(&self) -> f64 {
        let end = self.ended_at.unwrap_or_else(Utc::now);
        (end - self.started_at).num_milliseconds() as f64 / 1000.0
    }

    pub fn formatted_duration(&self) -> String {
        let secs = self.duration_secs();
        if secs < 60.0 {
            format!("{:.0}s", secs)
        } else if secs < 3600.0 {
            format!("{}m {}s", (secs / 60.0) as i32, (secs % 60.0) as i32)
        } else {
            let hours = (secs / 3600.0) as i32;
            let mins = ((secs % 3600.0) / 60.0) as i32;
            format!("{}h {}m", hours, mins)
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionSummary {
    pub id: Uuid,
    pub name: String,
    pub status: SessionStatus,
    pub agent_type: AgentType,
    pub started_at: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ended_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub metrics: SessionMetrics,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub working_directory: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub process_id: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_message: Option<String>,
    #[serde(default)]
    pub is_external_process: bool,
}

impl From<&Session> for SessionSummary {
    fn from(session: &Session) -> Self {
        Self {
            id: session.id,
            name: session.name.clone(),
            status: session.status.clone(),
            agent_type: session.agent_type.clone(),
            started_at: session.started_at,
            ended_at: session.ended_at,
            metrics: session.metrics.clone(),
            working_directory: session.working_directory.clone(),
            process_id: session.process_id,
            error_message: session.error_message.clone(),
            is_external_process: session.is_external_process,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum SessionStatus {
    Running,
    Paused,
    Completed,
    Failed,
    Waiting,
    Cancelled,
}

impl SessionStatus {
    pub fn color(&self) -> &'static str {
        match self {
            Self::Running => "green",
            Self::Paused => "yellow",
            Self::Completed => "blue",
            Self::Failed => "red",
            Self::Waiting => "orange",
            Self::Cancelled => "gray",
        }
    }

    pub fn icon(&self) -> &'static str {
        match self {
            Self::Running => "play-circle",
            Self::Paused => "pause-circle",
            Self::Completed => "check-circle",
            Self::Failed => "x-circle",
            Self::Waiting => "clock",
            Self::Cancelled => "stop-circle",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum AgentType {
    #[serde(alias = "Claude Code", alias = "claudeCode")]
    ClaudeCode,
    #[serde(alias = "codex")]
    Codex,
    #[serde(alias = "Custom Agent", alias = "custom")]
    Custom,
}

impl AgentType {
    pub fn display_name(&self) -> &'static str {
        match self {
            Self::ClaudeCode => "Claude Code",
            Self::Codex => "Codex",
            Self::Custom => "Custom Agent",
        }
    }

    pub fn icon(&self) -> &'static str {
        match self {
            Self::ClaudeCode => "brain",
            Self::Codex => "code",
            Self::Custom => "cpu",
        }
    }

    pub fn color(&self) -> &'static str {
        match self {
            Self::ClaudeCode => "purple",
            Self::Codex => "green",
            Self::Custom => "blue",
        }
    }

    pub fn default_executable(&self) -> &'static str {
        match self {
            Self::ClaudeCode => "claude",
            Self::Codex => "codex",
            Self::Custom => "agent",
        }
    }

    pub fn executable_names(&self) -> Vec<&'static str> {
        match self {
            Self::ClaudeCode => vec!["claude", "claude-code", "claude_code"],
            Self::Codex => vec!["codex", "openai-codex"],
            Self::Custom => vec!["agent"],
        }
    }

    pub fn default_args(&self) -> Vec<&'static str> {
        match self {
            Self::ClaudeCode => vec![],
            Self::Codex => vec!["--no-alt-screen"],
            Self::Custom => vec![],
        }
    }

    pub fn is_terminal_based(&self) -> bool {
        match self {
            Self::ClaudeCode | Self::Codex => true,
            Self::Custom => false,
        }
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionMetrics {
    #[serde(default)]
    pub total_tokens: i64,
    #[serde(default)]
    pub input_tokens: i64,
    #[serde(default)]
    pub output_tokens: i64,
    #[serde(default)]
    pub tool_call_count: i32,
    #[serde(default)]
    pub error_count: i32,
    #[serde(default)]
    pub api_calls: i32,
    #[serde(default)]
    pub cache_read_tokens: i64,
    #[serde(default)]
    pub cache_write_tokens: i64,
}

impl SessionMetrics {
    pub fn formatted_tokens(&self) -> String {
        let total = self.total_tokens;
        if total >= 1_000_000 {
            format!("{:.1}M", total as f64 / 1_000_000.0)
        } else if total >= 1_000 {
            format!("{:.1}K", total as f64 / 1_000.0)
        } else {
            total.to_string()
        }
    }
}
